docs/madr/decisions/046-meshservice-policy-matching.md

```diff
@@ -0,0 +1,313 @@
+* use exactly one of `labels` or `name`/`namespace`
+* no new top level fields besides `labels`
+
+### `targetRef`
+
+The policy is applied
+on both the Envoy outbound listeners and clusters
+that correspond to the `MeshService`.
+
+References for `kind: MeshService` have the following structure:
+
+```yaml
+---
+kind: MeshService
+spec:
+  ports:
+  - port: 80
+    name: http
+---
+targetRef:
+  kind: MeshService
+  namespace: backend
+  name: backend # only name/namespace or labels
+  labels: # only name/namespace or labels
+    kuma.io/zone: east
+  sectionName: http
+```
+
+The `name` of a `MeshService` ref always refers to the name of one, real Kubernetes
+object in the zone of the policy.
+
+Exactly one of either `targetRef.name` or `targetRef.labels` is **required**.
+
+#### `sectionName`
+
+If `sectionName` is set in `targetRef.kind` with `kind: MeshService`,
+it refers to an entry in `ports` by name and only traffic to that port is affected.
+
+In `backendRefs`, `port` is required. There is no `sectionName`, in accordance
+with Gateway API HTTPRoute.
+
+#### `name`
+
+If the policy ref sets `name`, it refers to one `MeshService` in the *local
+zone*.
+
+On a k8s zone:
+* without `namespace`, the policy refers to `MeshServices` in the same
+  namespace as the policy.
+* with `namespace`, only that namespace is searched for matching `MeshServices`
+
+On a universal zone, `namespace` is not valid.
+
+Note that this means it's possible to refer to synced `MeshServices` via their
+transformed `<service>-<hash-suffix>` name.
+
+#### `labels`
+
+If `name` is not set, `labels` must be set and `namespace` cannot be set!
+
+Via `labels`, groups of `MeshServices` can be matched, for example by:
+
+* `kuma.io/display-name`
+* `kuma.io/zone`
+
+With `labels`, we don't assume anything about the namespace of the object. The
+namespace _must_ be selected using the `k8s.kuma.io/namespace` label.
+
+This also means more verbosity for the headless service usecase.
+It means we don't infer any similarity between the local namespace
+and the other zone's namespace.
+
+#### Examples
+
+These examples assume the `frontend` namespace and the zone named `local-zone`.
+
+##### `MeshService`
+
+Users can target `MeshServices` in the namespace of the policy
+directly with `name`:
+
+```yaml
+spec:
+ to:
+ - targetRef:
+     kind: MeshService
+     name: backend
+```
+
+or by using `labels` without `name`. The below examples search _any namespace_.
+
+```yaml
+spec:
+ to:
+ - targetRef:
+     kind: MeshService
+     labels:
+       k8s.kuma.io/service-name: zk
+       k8s.kuma.io/namespace: zk-namespace # this must be set to target a namespace, even the local one
+       kuma.io/zone: local-zone # must also be set to target a local zone, otherwise all zones are targeted
+```
+
+```yaml
+spec:
+ to:
+ - targetRef:
+     kind: MeshService
+     labels:
+        kuma.io/display-name: backend
+        kuma.io/zone: other-zone
```

Comment:
With the producer and consumer model we're building up in: https://github.com/kumahq/kuma/pull/10148
It's not super clear how things work as now zone policies could be synced to other zones.

1. For example I have a backend service:

```yaml
kind: MeshService
metadata:
  name: my-backend
  namespace: my-backend-ns
spec:
  selector:
      dataplaneTags:
        app: my-backend
  ports:
    - port: 80
       targetPort: 8080
       name: http
       appProtocol: http
```

2. Intuitively I'd write this timeout policy that would affect all consumers of my service (in *ALL* zones):

```yaml
kind: MeshTimeout
metadata:
  name: my-backend
  namespace: my-backend-ns
spec:
  targetRef:
     kind: Mesh
  to:
    - targetRef:
          kind: MeshService
          name: my-backend
          port: 80 # (subquestion 1: here the alternative sectionName: http is valid too right?)
      default:
         timeout: 5s
```

3. Then I'd want to write a consumer policy in the same zone as:

```yaml
kind: MeshTimeout
metadata:
   namespace: my-frontend-ns
   name: override-my-backend
spec:
  targetRef:
     kind: Mesh
  to:
    - targetRef:
          kind: MeshService
          name: my-backend
          namespace: my-backend-ns
          port: 80 # (subquestion 1: here the alternative sectionName: http is valid too right?)
      default:
         timeout: 5s
```

4. And I'd write a consumer policy in a different zone as:

```yaml
kind: MeshTimeout
metadata:
   namespace: my-frontend-ns-in-another-zone
   name: override-my-backend
spec:
  targetRef:
     kind: Mesh
  to:
    - targetRef:
          kind: MeshService
          sectionName: http
          labels:
            kuma.io/display-name: my-backend
            k8s.kuma.io/namespace: my-backend-ns
            kuma.io/zone: my-other-k8s-cluster
      default:
         timeout: 10s
```

1 and 2 gets synced to remote zones. 3 and 4 don't. @lobkovilya is this correct?

Multiple questions come up to me:

- Do we find it acceptable that matching MeshService in the same zone vs cross zone is syntactically very different (Using labels for remote vs using name/ns for local)?
- With producer policies these gets synced to remote zones now. Is it still straight-forward to do the matching (do we use the synced labels to enhance the matcher)?

Reply:
I'm talking about [these rules](https://github.com/kumahq/kuma/pull/10148/files#diff-cb6c27d1ae4efdd51ae3730849991f19c4016f47fb91292fa349d9c0e0ea8a95R109-R111):
> The validation rules for namespace-scoped policy are the following:
> * to-policy when placed into the custom namespace must have the namespace and zone either in the top-level targetRef or in each to-item
> * from-policy when placed into the custom namespace must always have the namespace and zone in the top-level targetRef

Reply:
My messages above are not relevant, as we changed the policy on namespace MADR since then. 

> Do we find it acceptable that matching MeshService in the same zone vs cross zone is syntactically very different (Using labels for remote vs using name/ns for local)?

I think it's acceptable. You can still you name/ns for remote, but the name should be the one with hash suffix.  

> With producer policies these gets synced to remote zones now. Is it still straight-forward to do the matching (do we use the synced labels to enhance the matcher)?

In policy on namespace MADR we described `to[].targetRef.name` should be automatically updated to the name with hash suffix when synced to another zone.